### PR TITLE
db improvements

### PR DESCRIPTION
**Pull request by kderme**
https://github.com/input-output-hk/cardano-db-sync/issues/485
https://github.com/input-output-hk/cardano-db-sync/issues/484

### CODE REVIEW

File: schema/migration-3-0005-20210116.sql [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    -- Adding Indexes to Columns that reference other tables
+    -- https://www.postgresql.org/message-id/20040430163539.74079.qmail@web13805.mail.yahoo.com
+    -- Some of them are intentional commented out, because they already exist.
+    CREATE INDEX idx_delegation_pool_hash_id ON delegation(pool_hash_id) ;
+    CREATE INDEX idx_epoch_stake_pool_id ON epoch_stake(pool_id) ;
+    CREATE INDEX idx_orphaned_reward_pool_id ON orphaned_reward(pool_id) ;
+    CREATE INDEX idx_pool_owner_pool_hash_id ON pool_owner(pool_hash_id) ;
+    CREATE INDEX idx_pool_retire_hash_id ON pool_retire(hash_id) ;
+    --    CREATE INDEX idx_pool_update_hash_id ON pool_update(hash_id) ;
```
Comment (kderme) [inline schema/migration-3-0005-20210116.sql:16]: https://github.com/input-output-hk/cardano-db-sync/blob/b3f0f71fe5281295cfd4e464a62b398e3039f30c/schema/migration-3-0004-20200810.sql#L15

File: schema/migration-3-0005-20210116.sql [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    CREATE INDEX idx_reward_pool_id ON reward(pool_id) ;
+    CREATE INDEX idx_slot_leader_pool_hash_id ON slot_leader(pool_hash_id) ;
+    CREATE INDEX idx_block_slot_leader_id ON block(slot_leader_id) ;
+    --    CREATE INDEX idx_block_previous_id ON block(previous_id) ;
```
Comment (kderme) [inline schema/migration-3-0005-20210116.sql:20]: https://github.com/input-output-hk/cardano-db-sync/blob/b3f0f71fe5281295cfd4e464a62b398e3039f30c/schema/migration-3-0002-20200521.sql#L22

File: schema/migration-3-0005-20210116.sql [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    CREATE INDEX idx_epoch_param_block_id ON epoch_param(block_id) ;
+    CREATE INDEX idx_epoch_stake_block_id ON epoch_stake(block_id) ;
+    CREATE INDEX idx_orphaned_reward_block_id ON orphaned_reward(block_id) ;
+    CREATE INDEX idx_reward_block_id ON reward(block_id) ;
+    --    CREATE INDEX idx_tx_block_id ON tx(block_id) ;
```
Comment (kderme) [inline schema/migration-3-0005-20210116.sql:25]: https://github.com/input-output-hk/cardano-db-sync/blob/b3f0f71fe5281295cfd4e464a62b398e3039f30c/schema/migration-3-0002-20200521.sql#L25

File: schema/migration-3-0005-20210116.sql [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    CREATE INDEX idx_reserve_addr_id ON reserve(addr_id) ;
+    CREATE INDEX idx_reward_addr_id ON reward(addr_id) ;
+    CREATE INDEX idx_stake_deregistration_addr_id ON stake_deregistration(addr_id) ;
+    CREATE INDEX idx_stake_registration_addr_id ON stake_registration(addr_id) ;
+    CREATE INDEX idx_treasury_addr_id ON treasury(addr_id) ;
+    CREATE INDEX idx_tx_out_stake_address_id ON tx_out(stake_address_id) ;
+    CREATE INDEX idx_withdrawal_addr_id ON withdrawal(addr_id) ;
+    CREATE INDEX idx_ma_tx_out_tx_out_id ON ma_tx_out(tx_out_id) ;
+    CREATE INDEX idx_pool_update_meta_id ON pool_update(meta_id) ;
+    CREATE INDEX idx_pool_relay_update_id ON pool_relay(update_id) ;
```
Comment (kderme) [inline schema/migration-3-0005-20210116.sql:55]: 45 indexes,one for each foreign keys we have (minus 3 which already exist)

File: cardano-db-sync/src/Cardano/DbSync/LedgerState.hs [hunk 5]
```diff
@@ -143,16 +143,20 @@ applyBlock env (LedgerStateVar stateVar) blk =
...
   atomically $ writeTVar stateVar ledger
   case synced of
-    SyncFollowing -> saveState                      -- If following, save every state.
+    SyncFollowing -> saveState                          -- If following, save every state.
     SyncLagging
-      | unSlotNo slot == 0 -> pure ()               -- Genesis and the first EBB are weird so do not store them.
-      | block `mod` 2000 == 0 -> saveState          -- Only save state ocassionally.
+      | unSlotNo slot == 0 -> pure ()                   -- Genesis and the first EBB are weird so do not store them.
+      | block `mod` 2000 == 0 -> saveState              -- Only save state ocassionally.
+      | isJust (lssEpochUpdate snapshot) -> saveState   -- Epoch boundaries cost a lot, so we better save them
```
Comment (erikd) [inline cardano-db-sync/src/Cardano/DbSync/LedgerState.hs:154]: Made a very minor tweak here.

**Review by erikd — APPROVED** [review]

**Pull request merged**